---
title: "SARS-CoV-2 Non-human Shifts "
output: flexdashboard::flex_dashboard
runtime: shiny
---

```{r setup, include=FALSE}
library(shiny)
library(tidyverse)
library(flexdashboard)
library(ape)
library(phangorn)
library(DECIPHER)
library(tictoc)
library(ggtree)
library(plotly)
```

```{r global, include=FALSE}
KnownSeqs <- read.dna("all_aligned.fasta", format = "fasta")

KnownSeqsMeta <- read.table("metadata.csv", header = T, sep = ',')
```

Column {.sidebar}
-----------------------------------------------------------------------

Upload your data here.

```{r}
fileInput("sample", label = "Sequence data:", buttonLabel = "Browse...", placeholder = "No file selected")

selectInput("species", label = "Select organism:",
            choices = c("", "All", "canis_lupus", "felis_catus", "human", "mink",
                        "panthera_leo", "panthera_tigris_jacksoni"), selected = "")

selectInput("location", label = "Select location:",
            choices = c("", "All", "China", "Netherlands", "Poland", "Wisconsin-USA", "Unknown"),
            selected = "")
```

Column
-----------------------------------------------------------------------

### SARS-CoV-2 Tree

```{r}
# Defining dataset to use
renderTable({
#renderPlot({
  if(input$species == "All" & input$location == "All"){
    DesiredSeqs <- KnownSeqs
    DesiredIds <- KnownSeqsMeta$seqids
  }else if(input$species == "All" & input$location != "All"){
    DesiredIds <- KnownSeqsMeta %>% dplyr::filter(location == input$location) %>% select(seqid)
    DesiredIndex <- which(names(KnownSeqs) %in% DesiredIds)
    DesiredSeqs <- KnownSeqs[DesiredIndex]
  }else if(input$species != "All" & input$location == "All") {
    DesiredIds <- KnownSeqsMeta %>% dplyr::filter(species == input$species) %>% select(seqid)
    DesiredIndex <- which(names(KnownSeqs) %in% DesiredIds)
    DesiredSeqs <- KnownSeqs[DesiredIndex]
  }else if(length(input$species) & length(input$location)){
    DesiredIds <- KnownSeqsMeta %>% dplyr::filter(species == input$species, location == input$location) %>% select(seqid)
    DesiredIndex <- which(names(KnownSeqs) %in% DesiredIds)
    DesiredSeqs <- KnownSeqs[DesiredIndex]
  }
  DesiredIds
  
# Adding sample data
  #file <- input$sample
  #SampleSeq <- readDNAStringSet(file$datapath, format = "fasta")
  #DesiredSeqsConv <- DesiredSeqs %>% as.list %>% as.character %>%
   # lapply(., paste0, collapse = "") %>% unlist %>% DNAStringSet
  #aligned <- AlignProfiles(DesiredSeqsConv, SampleSeq)
  #names(as.DNAbin(aligned))
  
# Phylogenetic tree construction, uncomment when ready
# Initial tree
  #tic()
  #msa <- phyDat(as.DNAbin(aligned), type = "DNA")
  #dm  <- dist.ml(msa)
  #treeNJ <- NJ(dm)
  #ttime=toc
  
#ML tuning
#fit <- pml(treeNJ, data=aln)
#can also try ratchet/stochastic for rearrangement, but slower (better topology)
#fitGTR <- optim.plm(fit, model="GTR",optInv=TRUE,optGamma=TRUE,
#                    rearrangement="NNI",control=pml.control(trace=0))

# Plotting tree
  #t <- ggtree(treeNJ)
  #t
  #ggplotly(t)
})
```
